### DISCUSSION TITLE
Code Script in 0.39+?

### DISCUSSION THREAD

**Discussion by SamKook** [alejandroautalan/pygubu-designer#246] [Q&A] [answered]
I've been looking for a GUI for tkinter and after trying v 0.39.1, I went looking for alternatives after seeing it can only save the code on disk because I wanted to have code generated that I could copy-paste into my project since I'm almost certainly going to tinker with it as I code since I'm not a big planner.
After not really being happy with alternatives and having seen the hello world example in the wiki hinting at a code tab in the GUI in an earlier version, I had a look at the history only to discover that it was only removed recently(I assume it's recent at least from the version numbers, there is sadly no release dates in the history file) so I installed 0.38 and the Code Script template in the code section is exactly what I want, with a preview I can select what I want to copy-paste.

So, I was wondering if it was only removed temporarily because of the new GUI redesign(would be nice to have the possibility to have the components horizontally too, the search is great but I like the component layout of the older version better since it can be more compact)  or if the plan was to remove it for good, which would be really sad, and why was that decision taken if that is indeed the case.

**Comment by alejandroautalan** [comment, marked as answer]
Hello @SamKook, thanks for trying pygubu.
Sorry for the inconveniences.

The code tab was removed because the designer now can generate from 2 to 3 files depending on the project template.
On the project properties you can configure where those files will be generated. In your case, you can configure a temporary folder and copy the required code from there.

The components bar was changed for two reasons.
- I can easily change the palette components. This is useful when the project has custom widgets: they are loaded and displayed (per project) in the components palette.
- Some users have very large screens, so I created the docking system to be able to position the tools where they like best.
- Another reason is that the number of components will be increasing with time, so I think the treeview is the best approach (at least is more easy to code)

This changes are not temporary. I will try to update the docs.

Regards.
Alejandro A.

**Comment by SamKook** [comment]
Thanks for the reply.

Having to open a file every time is quite the hassle for the way I want to work since it'll be frequent. In a perfect world I wanted something that could switch seamlessly between code and design view with the ability to edit either and have it reflected on the other, but that doesn't seem to exist for tkinter. At least having the code generation in the same app was closer to what I wished for.
Tabs could have been used for the different files, but with a multi file output, most people would want it saved I guess, would have been nice to have the option to choose though.

As for the component change, it's basically because I have a very large screen(34" 21:9 + 29" 21x9 above) that I want a more compact UI, lol.
Having the ability to have all my windows in view at all time is great, but that usually means one monitor is split into 2 or 3 sections, which means they end up being narrower than they are tall so vertical elements are harder to position properly rather than horizontal ones.
The docking is great but the problem is that the component content won't spread horizontally so there's not many different positions that make a ton of sense unless you want to constantly scroll.
I would have expected the components could be filled dynamically from a list system you can make your own standard for regardless of what they are but I'm not very familiar with python GUIs so you'd know better what's best and something easy to maintain and modify is very important.

Anyway, after messing around with it for an hour or so yesterday, I basically came to the conclusion that I'm not really saving time using the GUI but it helped me understand better how laying stuff out works so I'm very grateful for that at least.

One big issue I ran into when designing something was the inability to specify only one side for the padding. In code, you can specify (5,0) for pady if you only want 5 padding on top and nothing on the bottom which is essential for me since I like a more compact UI so I'd have to modify the code quite a bit after generating it and wouldn't be able to preview what it actually looks like during the design phase.

Another one was that there doesn't seem to be a cascade component for the menu so I'm not sure how we're even supposed to create a menu, but that might very well be that I missed something(haven't used tkinter a lot so not too familiar with it) and I haven't really looked that much into how since it wasn't making much sense for me to use it at that point anyway.

**Reply by alejandroautalan** [reply]
Hello @SamKook,
Thank you for your comments, I will take them into account in the future.

> One big issue I ran into when designing something was the inability to specify only one side for the padding. In code, you can specify (5,0) for pady if you only want 5 padding on top and nothing on the bottom which is essential for me since I like a more compact UI so I'd have to modify the code quite a bit after generating it and wouldn't be able to preview what it actually looks like during the design phase.

You can do that entering values separated by space.

[padding.webm](https://github.com/alejandroautalan/pygubu-designer/assets/3482471/51ce39a5-6340-41d2-91ec-5dc92a66f412)

> Another one was that there doesn't seem to be a cascade component for the menu so I'm not sure how we're even supposed to create a menu, but that might very well be that I missed something(haven't used tkinter a lot so not too familiar with it) and I haven't really looked that much into how since it wasn't making much sense for me to use it at that point anyway.

Maybe is not intuitive, but you can create cascade menus very easily.

[cascade-menu.webm](https://github.com/alejandroautalan/pygubu-designer/assets/3482471/d90a86e9-8a3b-4f9d-9404-c01ab8aa4e79)

You can find a more explained video [here](https://www.youtube.com/watch?v=PPbxbj896DI)

Regards
Alejandro A.

**Reply by SamKook** [reply]
I did not think to try "5 5", I tried "(5, 5)" and "5, 5", good to know it's possible.


That's pretty much what I tried for the menu except I didn't touch the tearoff option, but even if I do, I get this:
![pygubu_menu](https://github.com/alejandroautalan/pygubu-designer/assets/84121956/d92d91d3-509e-4f48-867d-9ba85e203afe)
